### DISCUSSION TITLE
New path using separate service over offchain-worker.

### DISCUSSION THREAD

**Discussion by drewstone** [hicommonwealth/substrate-rainbow-bridge#10] [Ideas]
Should the relayer or the offchain worker be "dumb"?

**Comment by drewstone** [comment]
If the relayer is the "dumb" service then we will require it to:
- receive an rlp header and generate the DAG if it hasn't already, otherwise generate a proof and send it back
- regenerate the next DAG at some threshold
- strictly serve proofs to the offchain worker

The offchain worker will:
- query INFURA for the next block it needs (potentially many next blocks it needs), receive the rlp headers and encode them
- send HTTP POST to the relayer asking for the proof
- try to submit proof txes

**Comment by shekohex** [comment]
So after taking some time to think more about it, here is my vision or my plan to build this:

We will make the Relayer as dumb as possible, it makes it simple to implement and to verify, and everything else should happen in the off-chain worker.

As you said, Drew, I'm not going to add much but I will just make it a bit organized and simple to follow:

### From an Off-chain Worker point of view: 

The off-chain worker will use the off-chain storage to store a Queue of RLP Encoded `BlockHeaders`, that makes sure we will never going to miss or lose any data.

#### Steps for OCW:

Check if there are any `BlockHeaders` are still in the Queue, and there are two possible branches here:
1. There are **no** pending headers in the queue:
      in this case, send a query to INFRUA and get the RLP Encoded header from it, and then `lock` for `LOCK_TIME` and send it to the Relayer and wait for the response.
     the response is either the service is busy (i.e doing a DAG dataset generation, which may take several minutes) or return the proofs.
     in the first case, we simply add this `BlockHeader` RLP encoded (`Vec<u8>`) to the Queue.
2. There are pending headers in the queue:
      in this case, we get the first header (the queue is FIFO), and we `lock` for `LOCK_TIME` and send it to the Relayer and wait for the response back, if the Relayer is busy; re-queue that header again (in front of the queue) and fetch another `BlockHeader` from INFRUA and enqueue it (add it to the back of the queue).

Notes: 
`LOCK_TIME`: should be implementation-specific (default to `5 * 60_000` ms)

### From a Relayer point of view:

The Relayer should act as a simple yet "dumb" HTTP server with only a few (currently one) endpoints. 

it should only handle these cases:
1. Generating DAG (Cache and Datasets).
2. Generating Proofs.

The endpoint should look something like this:

```sh
POST /proofs

{
    "rlp": "RLP_ENCODED_BLOCK_HEADER_IN_HEX_FORMAT"
}
```

#### Steps for Relayer:

Receiving an HTTP to (`/proofs`) with the RLP Encoded `BlockHeader` the server should do the following:

1. Decode the Hex value back to `Vec<u8>`.
2. Decode the `Vec<u8>` back to the `BlockHeader` struct.
3. Calculate the `epoch` form the current `block_number` (`epoch = block_number / 30_000`).
4. Load the DAG (Cache, Dataset) for that epoch:
    a. if this data exists; start by calculating the  `indices` and building up the `MerkleTree` and generating proofs for these `indices`  and go to step 5.
    b. if this data does **not** exist, go to step 6.
5. Send back the proofs in a similar format to the one in the test fixtures (example: [2.json](https://github.com/edgeware-builders/substrate-eth-light-client/blob/7416ccda5b2648c464c851f7f21de7bc515b07da/pallets/worker/data/2.json)).
6. Start by generating the DAG (Cache, Dataset) and saving them in the filesystem (see next) and return `423 Locked` HTTP Response.
7. Delete the Old DAG files for the previous epoch (if exist).

##### Saving DAG Files:

Simply the DAG files are saved in a path like this.

Cache: `./data/dag/{epoch}/cache.bin`
Dataset: `./data/dag/{epoch}/dataset.bin`

#### Simple Diagram for the final Architecture:
![ETH-Relayer](https://user-images.githubusercontent.com/14620076/104074445-10342180-5208-11eb-81d3-d39991686944.jpg)

Please comment/reply with any questions or thoughts.

**Comment by shekohex** [comment]
Hey @drewstone 

The work has started here: https://github.com/edgeware-builders/ethash-relyer
and almost done :)
